dali/operators/signal/fft/spectrogram.cc

```diff
@@ -0,0 +1,276 @@
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/data/views.h"
+
+#define SPECTROGRAM_SUPPORTED_NDIMS (1, 2)
+
+namespace dali {
+
+DALI_SCHEMA(Spectrogram)
+  .DocStr(R"code(Produces a spectrogram from a 1D signal (e.g. audio). Input data is expected
+to be single channel (1D shape `(time)`) or multi channel in planar layout (channel, time) 32 bit
+float tensor.)code")
+  .NumInput(1)
+  .NumOutput(1)
+  .AddOptionalArg("nfft",
+    R"code(Size of the FFT. The number of bins created in the output is `nfft // 2 + 1`
+(positive part of the spectrum only).)code",
+    -1)
+  .AddOptionalArg("window_length",
+    R"code(Window size (in number of samples))code",
+    512)
+  .AddOptionalArg("window_step",
+    R"code(Step betweeen the STFT windows (in number of samples))code",
+    256)
+  .AddOptionalArg("window_fn",
+    R"code(Samples of the window function that will be multiplied to each extracted window when
+  calculating the STFT. If provided it should be a list of floating point numbers of size
+  `window_length`. If not provided, a Hann window will be used.)code",
+    std::vector<float>{})
+  .AddOptionalArg("power",
+    R"code(Exponent of the magnitude of the spectrum. Supported values are 1 for energy and 2 for
+power.)code",
+    2)
+  .AddOptionalArg("center_windows",
+    R"code(Indicates whether extracted windows should be padded so that window function is centered
+at multiples of `window_step`.)code",
+    true)
+  .AddOptionalArg("reflect_padding",
+    R"code(Indicates the padding policy when sampling outside the bounds of the signal. If set to
+true, the signal is mirrored with respecto to the boundary, otherwise the signal is padded with
+zeros.)code",
+    true);
+
+template <int Dims>
+struct SpectrogramImplCpu : detail::OpImplBase<CPUBackend> {
+  using OutputType = float;
+  using InputType = float;
+
+  static constexpr int InputDims = Dims;
+  using WindowKernel =
+    kernels::signal::window::ExtractWindowsCpu<InputType, InputType, InputDims>;
+
+  static constexpr int WindowsDims = Dims+1;
+  using FftKernel = kernels::signal::fft::Fft1DCpu<OutputType, InputType, WindowsDims>;
+
+  explicit SpectrogramImplCpu(const OpSpec & spec);
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<CPUBackend> &ws) override;
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  int nfft_ = -1;
+  int64_t window_length_ = -1;
+  int64_t window_step_ = -1;
+  int power_ = -1;
+  bool reflect_padding_ = true;
+  std::vector<float> window_fn_;
+  int64_t window_center_ = -1;
+
+  kernels::KernelManager kmgr_window_;
+  kernels::signal::window::ExtractWindowsArgs window_args_;
+
+  std::vector<OutputDesc> window_out_desc_;
+  std::vector<Tensor<CPUBackend>> window_out_;
+
+  kernels::KernelManager kmgr_fft_;
+  kernels::signal::fft::FftArgs fft_args_;
+};
+
+namespace {
+  void FillExtractWindowsArgs(kernels::signal::window::ExtractWindowsArgs& args,
```

Comment:
There's very little logic in this function and it's used only once - maybe you can just use braced initializer at call site instead?

Reply:
Removed